lib/mayaUsd/ufe/UsdUndoInsertChildCommand.cpp

```diff
@@ -0,0 +1,161 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "UsdUndoInsertChildCommand.h"
+#include "Utils.h"
+#include "private/InPathChange.h"
+
+#include <ufe/scene.h>
+#include <ufe/sceneNotification.h>
+#include <ufe/log.h>
+#ifdef UFE_V2_FEATURES_AVAILABLE
+#define UFE_ENABLE_ASSERTS
+#include <ufe/ufeAssert.h>
+#else
+#include <cassert>
+#endif
+
+#include <pxr/usd/usd/stage.h>
+#include <pxr/usd/usd/editContext.h>
+#include <pxr/usd/sdf/copyUtils.h>
+#include <pxr/base/tf/token.h>
+
+MAYAUSD_NS_DEF {
+namespace ufe {
+
+UsdUndoInsertChildCommand::UsdUndoInsertChildCommand(
+    const UsdSceneItem::Ptr& parent,
+    const UsdSceneItem::Ptr& child,
+    const UsdSceneItem::Ptr& pos
+) : Ufe::UndoableCommand()
+{
+    // First, check if we need to rename the child.
+    auto childName = uniqueChildName(parent, child->path());
+
+    // Set up all paths to perform the reparent.
+    auto childPrim  = child->prim();
+    fStage      = childPrim.GetStage();
+    fUfeSrcItem = child;
+    fUsdSrcPath = childPrim.GetPath();
+    fUfeDstPath = parent->path() + childName;
+    fUsdDstPath = parent->prim().GetPath().AppendChild(TfToken(childName));
+
+    fLayer = defPrimSpecLayer(childPrim);
+    if (!fLayer) {
+        std::string err = TfStringPrintf("No prim found at %s", childPrim.GetPath().GetString().c_str());
+        throw std::runtime_error(err.c_str());
+    }
+}
+
+UsdUndoInsertChildCommand::~UsdUndoInsertChildCommand()
+{
+}
+
+/*static*/
+UsdUndoInsertChildCommand::Ptr UsdUndoInsertChildCommand::create(
+    const UsdSceneItem::Ptr& parent,
+    const UsdSceneItem::Ptr& child,
+    const UsdSceneItem::Ptr& pos
+)
+{
+    return std::make_shared<UsdUndoInsertChildCommand>(parent, child, pos);
+}
+
+bool UsdUndoInsertChildCommand::insertChildRedo()
+{
+    // See comments in UsdUndoRenameCommand.cpp.
+    bool status = SdfCopySpec(fLayer, fUsdSrcPath, fLayer, fUsdDstPath);
```

Comment:
If this looks very much like the rename implementation, it's because it is: the same strategy is used.  We copy the USD source hierarchy into the destination (to reparent), then inactivate the source.

We are currently investigating alternative implementations for rename and reparent.